### ISSUE TITLE
Smoothing off the rough edges

### ISSUE THREAD

**Issue by mbland** [18F/team_api#2]
Now that the `.about.yml` => https://team-api.18f.gov/public/api/ => https://18f.gsa.gov/dashboard/ pipeline is set up and working, there are a number of "gotchas" that it would be good to guard against to avoid confusion and frustration:

- [x] There's actually a bit of a race condition right now. When an `.about.yml` update comes in, the site isn't rebuilt until _after_ the update is committed, causing the webhook to fire before the new data is available to the Dashboard. (I've been manually re-firing webhooks by hand.) There's a _really_ easy fix for this; I'll try to file a PR this weekend.
- [ ] We need to add extra schema validation to [`about_yml_generate`](https://github.com/18F/about_yml) beyond what's in the `lib/about_yml/schema.json` file. Namely, we should check that all `team:` members are able to be cross-referenced with what's currently in the Team API, and provide useful hints if something doesn't line up.
  - i.e. Team member information must be up-to-date in 18F/team-api before new 18F team members can be added to an `.about.yml` file.
- [ ] We need better ways of keeping team member info up-to-date, to migrate from 18F/data-private to 18F/team-api, better motivate folks to keep their team member data up-to-date, etc. etc.
- [ ] We need to settle some of the schema issues, and the use `about_yml_validate` to clean up existing `.about.yml` files. The [`about_yml_scrape`](https://github.com/18F/about_yml/blob/master/bin/about_yml_scrape) program will actually grab _all_ 18F `.about.yml` files and validate them in one gulp.
- [ ] The `.about.yml` format _mostly_ lines up with the current format for 18F/data-private `projects/*.yml` files, but there are some differences. The sooner we can migrate from the data-private format to `.about.yml`, the sooner we can update the Dashboard templates and get rid of some data munging code from the Dashboard's `_plugins/dashboard.rb`.
- [ ] The deployment flow needs to be documented, I'm not sure where. But at a high level, here it is:
  - Change to an `.about.yml` file committed.
  - A webhook for a `push` event is fired for the 18F GitHub organization to https://team-api.18f.gov/pshb.
    - This webhook is implemented in 18F/team-api, via the `deploy/hookshot.js` script.
    - It's launched using [Fabric](http://www.fabfile.org/) by running `fab start` in the `deploy` directory.
    - On the team-api.18f.gov server, the `fab start` command launches the `hookshot.js` script as a [Forever](https://github.com/foreverjs/forever) job; you can run `forever list` to see it on the server.
    - The server is the same as the Hub; 18F/hub contains the [team-api nginx configuration](https://github.com/18F/hub/blob/master/deploy/etc/nginx/vhosts/team-api.conf).
    - The log on the server is `/home/ubuntu/team-api.log`.
  - The webhook parses the update and creates an automatic commit from the new `about.yml` data.
  - The automatic commit fires a webhook for https://18f.gsa.gov/dashboard/team-api-update.
  - The Dashboard's [`team-api-update.js` server](https://github.com/18F/dashboard/blob/staging/deploy/team-api-update.js) parses the webhook, pulls the new https://team-api.18f.gov/public/api/projects/ data, and builds https://staging.18f.us/dashboard/.
    - The setup is similar to the Team API server, except that the job is launched with the [`team-api-update-deploy.sh` script](https://github.com/18F/dashboard/blob/staging/deploy/team-api-update-deploy.sh).
  - If the staging build is successful, it then builds https://18f.gsa.gov/dashboard/.

Perhaps there's more, and perhaps we need separate issues. I just wanted to get several of them down here so that we all can get an idea of what's going on, what the gotchas are, what we need to do to fix them, etc. Feel free to add more, discuss, break these out into new issues, and so on.

Another motivation: I want the bus number for this to be strictly greater than one, especially while I'm on vacation.

But, hey, the thing pretty much works, and that's kinda cool. :smile:

cc: @arowla @gboone @melodykramer @jeremiak @jmcarp @monfresh @wslack @leahbannon @afeld @catherinedevlin

**Comment by leahbannon** [comment]
Wow, what a list. Also, a couple questions:

- [ ] For `blog`, we used to just put in the tag used on our blog. Is that still how that field works?
- [ ] For `contact:`, the description says to post an email address. Formerly, it would link to the github repo if if you did something like `- 18F/FEC/issues` -- does that still work?

**Comment by mbland** [comment]
So I marked off one item here: The race condition was fixed in 18F/team-api#105.

@leahbannon: Not sure about the `blog:` field. Perhaps @gboone could say more? And no, the `contact:` field only takes email addresses now. The `links:` field can accommodate links to repos. But, if that old behavior is really desired, file an issue and we can consider restoring it.

**Comment by gboone** [comment]
IIRC @leahbannon is correct about the `blog` field. [Every tag listed under `blog` will link out to the archive at `18f.gsa.gov/tags/{{tag}}`](https://github.com/18F/dashboard/blob/staging/_layouts/project.html#L25-L38)

As for `contact`: Dashboard [will still support that behavior, if the Team API allows it](https://github.com/18F/dashboard/blob/staging/_layouts/project.html#L14-L24) Though I'm inclined to think `contact` should be a way people can contact 18F about a project without necessarily having a GitHub account.

**Comment by gboone** [comment]
> We need better ways of keeping team member info up-to-date, to migrate from 18F/data-private to 18F/team-api, better motivate folks to keep their team member data up-to-date, etc. etc.
Maybe a web form somewhere in the Hub that POSTs to the GitHub API?

**Comment by mbland** [comment]
_Ooohh_, I like that idea, @gboone. I think that's totally doable. _If_ I had a dedicated frontend dev... Or maybe it's time I grew up and gave it a shot?

**Comment by mbland** [comment]
And, looking at the code, the `about_yml` gem doesn't actually validate that it's an email address. In other words, it'll just pass the plain text through, which means the Dashboard should get it already.

**Comment by mbland** [comment]
@gboone One for for you: What do you think about pushing that logic into the Team API, versus having it in the Dashboard template? In other words, the Team API could just provide an anchor text, URL pair, munged who-knows-how from who-cares-where, and the Dashboard just consumes that?

**Comment by gboone** [comment]
I like that idea. Right now if someone had a google form they used for `contact` it would fill in `https://github.com/<url-to-the-google-form`/ Having that data from in clean from the start would be more helpful than layering more and more complexity on that template.

**Comment by mbland** [comment]
Mind filing an issue on this repo? Happy to do it if you don't have the bandwidth.

**Comment by leahbannon** [comment]
Sorry, I'm confused. @mbland can you confirm @gboone's comments here https://github.com/18F/team_api/issues/2#issuecomment-140762925 ?

**Comment by mbland** [comment]
So if I'm understanding @gboone correctly, as it stands now, the `blog:` field works the way it always has, and there are no plans to change that. (It's passed straight through from the `.about.yml` to the Team API output and into the Dashboard.)

But in the case of `contact:`, we're considering making _that_ field a bit more flexible by having the entries change to (url, anchor text) pairs, so the Dashboard can just render those entries without its own special logic. For example, this:

```yaml
contact:
- michael.bland@gsa.gov
- 18F/team_api/issues
```

Would become this:

```yaml
contact:
- url: mailto:michael.bland@gsa.gov
  text: Mike Bland
- url: https://github.com/18F/team_api/issues
  text: 18F/team_api GitHub issues
```

In other words, `contact:` would be structured just like the `links:` field, but be treated slightly differently.

**Comment by leahbannon** [comment]
ah, that sounds cool :+1:

**Comment by mbland** [comment]
I've broken the outstanding issues into #23, #24, and #25.
